backend/src/mock/connection_constraint_checker.rs

```diff
@@ -151,43 +160,74 @@ pub struct ConnectionConstraintChecker<'a, F: FieldElement> {
     /// Checks all connections.
-    pub fn check(&self) {
-        for connection in self.connections {
-            self.check_connection(connection);
-        }
+    pub fn check(&self) -> Result<(), FailingConnectionConstraints<'a, F>> {
+        let errors = self
+            .connections
+            .iter()
+            .filter_map(|connection| self.check_connection(connection).err())
+            .collect::<Vec<_>>();
+
+        (!errors.is_empty())
+            .then(|| {
+                let error = FailingConnectionConstraints {
+                    connection_count: self.connections.len(),
+                    errors,
+                };
+                log::error!("{}", error);
+                Err(error)
+            })
+            .unwrap_or(Ok(()))
     }
 
     /// Checks a single connection.
-    fn check_connection(&self, connection: &Connection<F>) {
+    fn check_connection(
+        &self,
+        connection: &'a Connection<F>,
+    ) -> Result<(), FailingConnectionConstraint<'a, F>> {
         let caller_set = self.selected_tuples(&connection.caller(), &connection.left);
         let callee_set = self.selected_tuples(&connection.callee(), &connection.right);
 
         match connection.kind {
             ConnectionKind::Lookup => {
-                for tuple in caller_set {
-                    assert!(
-                        callee_set.contains(&tuple),
-                        "Lookup failed: {tuple:?} not found in callee",
-                    );
+                // Check if $caller \subseteq callee$.
+                let caller_set = caller_set.into_iter().collect::<HashSet<_>>();
+                let callee_set = callee_set.into_iter().collect::<HashSet<_>>();
+                let not_in_callee = caller_set
+                    .difference(&callee_set)
+                    .cloned()
+                    .collect::<Vec<_>>();
+                if !not_in_callee.is_empty() {
+                    Err(FailingConnectionConstraint {
+                        connection,
+                        not_in_callee,
+                        not_in_caller: Vec::new(),
+                    })
+                } else {
+                    Ok(())
                 }
             }
             ConnectionKind::Permutation => {
-                assert_eq!(
-                    caller_set.len(),
-                    callee_set.len(),
-                    "Permutation failed: caller and callee have different sizes"
-                );
-                for tuple in &caller_set {
-                    assert!(
-                        callee_set.contains(tuple),
-                        "Permutation failed: {tuple:?} not found in callee",
-                    );
-                }
-                for tuple in &callee_set {
-                    assert!(
-                        caller_set.contains(tuple),
-                        "Permutation failed: {tuple:?} not found in caller",
-                    );
```

Comment:
This was actually buggy before! It checked only set equality, not multi-set equality.